service/query/evm.go

```diff
@@ -0,0 +1,239 @@
+			topicsS = append(topicsS, topic.Hex())
+		}
+		rLog := client.EVMLogReply{
+			TransactionHash: log.TxHash.Hex(),
+			BlockHeight:     strconv.Itoa(int(log.BlockNumber)),
+			BlockHash:       log.BlockHash.Hex(),
+			Address:         keys.Address(log.Address.Bytes()),
+			Data:            ethcmn.Bytes2Hex(log.Data),
+			Topics:          topicsS,
+		}
+		reply.Logs = append(reply.Logs, rLog)
+	}
+	return nil
+}
+
+func (svc *Service) EVMAccount(args client.EVMAccountRequest, reply *client.EVMAccountReply) error {
+	acc, err := svc.accountKeeper.GetAccount(args.Address)
+	reply.Address = args.Address
+	reply.CodeHash = ethcmn.Bytes2Hex(ethcrypto.Keccak256(nil))
+	if err == nil {
+		reply.Balance = acc.Coins.String()
+		reply.Nonce = acc.Sequence
+	}
+	return nil
+}
+
+// EVMCall call smart contract code to get the result
+func (svc *Service) EVMCall(args client.SendTxRequest, reply *client.EVMCallReply) error {
+	height := svc.contracts.State.Version()
+	stateDB := action.NewCommitStateDB(svc.contracts, svc.accountKeeper, svc.logger)
+	bhash := stateDB.GetHeightHash(uint64(height))
+	stateDB.SetBlockHash(bhash)
+	// TODO: Change this
+	header := &abci.Header{
+		ChainID: "test-1",
+		Height:  height,
+		Time:    time.Now(),
+	}
+
+	var to *keys.Address
+	if len(args.To) != 0 {
+		to = &args.To
+	}
+	tx := action.RawTx{
+		Type: action.SC_EXECUTE,
+		Fee: action.Fee{
+			Price: args.GasPrice,
+			Gas:   args.Gas,
+		},
+	}
+
+	svc.logger.Debugf("Nonce: %d\n", args.Nonce)
+
+	evmTx := action.NewEVMTransaction(stateDB, header, args.From, to, args.Nonce, args.Amount.Value.BigInt(), args.Data, true)
+
+	// TODO: Move in some constant
+	timeout := 5 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	vmenv := evmTx.NewEVM()
+	// Wait for the context to be done and cancel the evm. Even if the
+	// EVM has finished, cancelling may be done (repeatedly)
+	go func() {
+		<-ctx.Done()
+		vmenv.Cancel()
+	}()
+
+	result, err := evmTx.Apply(vmenv, tx)
+	if vmenv.Cancelled() {
+		return fmt.Errorf("execution aborted (timeout = %v)", timeout)
+	}
+
+	if err != nil {
+		return fmt.Errorf("err: %w (supplied gas %d)", err, args.Gas)
+	}
+
+	svc.logger.Debugf("Result of call: %+v\n", result)
+
+	if result.Failed() {
+		return result.Err
+	}
+
+	// If the result contains a revert reason, try to unpack and return it.
+	if len(result.Revert()) > 0 {
+		return newRevertError(result)
+	}
+
+	*reply = client.EVMCallReply{
+		Result: ethcmn.Bytes2Hex(result.Return()),
+	}
+	return nil
+}
+
+// EVMEstimateGas call smart contract code to get the gas
+func (svc *Service) EVMEstimateGas(args client.SendTxRequest, reply *client.EVMEstimateGasReply) error {
+	height := svc.contracts.State.Version()
+	stateDB := action.NewCommitStateDB(svc.contracts, svc.accountKeeper, svc.logger)
+	bhash := stateDB.GetHeightHash(uint64(height))
+	stateDB.SetBlockHash(bhash)
+	// TODO: Change this
+	header := &abci.Header{
+		ChainID: "test-1",
```

Comment:
same here

Reply:
will fix, not a critical point for the simulation of vm.